### PR TITLE
Add support for custom tag attribute

### LINKED ISSUE #147: load css with media=all attribute
Hey there,

I initially posted the following issue to: https://github.com/webpack/css-loader/issues/294
Someone told me I would be better served posting it here instead. Do you think you can help me? 
Here is the issue:

I'd like to add the media=all attribute to the css files that webpack is generating:
image

The reason for this is because the css is not being used when I try to print my webpack-generated webpage. 

Thanks for the help!

### PR DESCRIPTION
**Pull request by felixmosh**
**What kind of change does this PR introduce?**
feature, allows to add additional attributes on the style / link element that the loader creates

**Did you add tests for your changes?**
No, tests
**If relevant, did you update the README?**
Update README
**Summary**

This PR allows to add id attribute or any custom attribute, in my case i have runtime css style processor that relies on custom attribute on the style tag that it should process. 

**Does this PR introduce a breaking change?**
no breaking changes

**Other information**

Fixes #147

### CODE REVIEW

**Comment by felixmosh** [comment]
@sokra can you review it?

**Comment by felixmosh** [comment]
someone is alive here?

File: addStyles.js [hunk 1]
```diff
@@ -26,6 +26,8 @@ module.exports = function(list, options) {
 	}
 
 	options = options || {};
+  options.tagAttrs = typeof options.tagAttrs === "object"? options.tagAttrs : {};
```
Comment (sokra) [inline addStyles.js]: spacing

**Comment by felixmosh** [comment]
Done, thanx

**Comment by sokra** [comment]
the spacing should be tabs instead of spaces

**Comment by felixmosh** [comment]
Fixed the spacing

File: README.md [hunk 2]
```diff
@@ -61,6 +61,10 @@ By default, the style-loader appends `<style>` elements to the end of the `<head
 
 If defined, the style-loader will re-use a single `<style>` element, instead of adding/removing individual elements for each required module. **Note:** this option is on by default in IE9, which has strict limitations on the number of style tags allowed on a page. You can enable or disable it with the singleton query parameter (`?singleton` or `?-singleton`).
 
+#### `tagAttrs`
+
+If defined, style-loader will attach given attributes with there values on `<style>` / `<link>` element.
```
Comment (SpaceK33z) [inline README.md]: there -> their

Comment (SpaceK33z) [inline README.md]: Can you also add what this option expects? Is it an array or object? And what do I exactly need to define in that?

**Comment by felixmosh** [comment]
All done...

**Comment by gruberjl** [comment]
This is exactly what I needed, thanks.

FYI: The docs say `require('style-loader?{tagAttrs:{id: 'style-tag-id'}}!style.scss');`. I think `style-tag-id` should be wrapped in double quotes. It did work what I used the options syntax in Webpack 2.

```
{
    loader: 'style-loader',
    options: {
        singleton: true,
        tagAttrs: {'amp-custom': ''}
    }
}
```

**Comment by felixmosh** [comment]
@gruberjl, you right, fixed it. 
Thanx.

This is exactly what i need and i waiting for the owner of the proj to approve it :\

**Comment by felixmosh** [comment]
Someone?

**Comment by d3viant0ne** [comment]
@felixmosh - Give me a minute, I'll see if I can find someone

File: addStyles.js [hunk 3]
```diff
@@ -128,18 +130,30 @@ function removeStyleElement(styleElement) {
...
 
 function createLinkElement(options) {
 	var linkElement = document.createElement("link");
-	linkElement.rel = "stylesheet";
+	options.tagAttrs.rel = "stylesheet";
+
+	attachTagAttrs(styleElement, options.tagAttrs);
 	insertStyleElement(options, linkElement);
 	return linkElement;
 }
```
Comment (d3viant0ne) [inline addStyles.js]: It might not hurt to throw a quick `// document this function` in there given it's an optional feature.

Comment (felixmosh) [inline addStyles.js]: Sorry, i didn`t understood, what does it means "to throw a quick // document this function"

**Comment by dsebastien** [comment]
This will be very nice to have as it should allow specifying a nonce to be used with a content security policy (CSP).

File: README.md [hunk 4]
```diff
@@ -61,7 +72,17 @@ By default, the style-loader appends `<style>` elements to the end of the `<head
 
 If defined, the style-loader will re-use a single `<style>` element, instead of adding/removing individual elements for each required module. **Note:** this option is on by default in IE9, which has strict limitations on the number of style tags allowed on a page. You can enable or disable it with the singleton query parameter (`?singleton` or `?-singleton`).
 
-## Recommended configuration
+#### `tagAttrs`
```
Comment (michael-ciniawsky) [inline README.md]: `tagAttrs` => `Attrs/Attributes`

File: addStyles.js [hunk 5]
```diff
@@ -26,6 +26,8 @@ module.exports = function(list, options) {
 	}
 
 	options = options || {};
+	options.tagAttrs = typeof options.tagAttrs === "object" ? options.tagAttrs : {};
```
Comment (michael-ciniawsky) [inline addStyles.js]: `tagAttrs` => `attrs/attributes`

File: addStyles.js [hunk 6]
```diff
@@ -128,18 +130,30 @@ function removeStyleElement(styleElement) {
...
-	linkElement.rel = "stylesheet";
+	options.tagAttrs.rel = "stylesheet";
+
+	attachTagAttrs(styleElement, options.tagAttrs);
 	insertStyleElement(options, linkElement);
 	return linkElement;
 }
 
+function attachTagAttrs(element, attrs) {
+	for(var key in attrs) {
```
Comment (michael-ciniawsky) [inline addStyles.js]:
```js
options.attrs.forEach((attr) => element.setAttribute(attr, attrs[attr]));
```

**Comment by michael-ciniawsky** [comment]
Related #119

File: addStyles.js (hunk 6 above)
Comment (felixmosh) [inline addStyles.js]: options.attrs is not an array but an key-value pair (object).
So forEach won`t work.

**Comment by d3viant0ne** [comment]
So I hate to tell you this but it looks like there was an issue updating this PR around Jan 9th ( oops on a rebase maybe ).

Point being, the CLA assistant is currently showing 4 contributors to this PR which isn't going to work.

Simplest ( and probably only ) way to fix this is to copy code changes to a new branch, Open a new PR, reference this one & then close it. We will finish this and get it merged as soon as we have a pull request that doesn't have the CLA issues.

**Comment by felixmosh** [comment]
So should i create a PR from scratch?

**Comment by d3viant0ne** [comment]
To prevent issues, it just needs to be a new branch. Everything ( all the changes from this PR ) can be committed in one hash.

We will reference this PR from the new one, so none of the review comments will be lost.

**Comment by madjam002** [comment]
@dsebastien The way that custom tag attributes have been implemented here is not suitable for implementing CSP nonce attributes. A nonce by definition is a secure random string that can only be used once [1], and by hard coding a nonce in your webpack config goes against this. By using the same nonce over and over again, malicious entities know what the nonce will be every time which defeats the point of CSP in the first place.

[1] https://en.wikipedia.org/wiki/Cryptographic_nonce

**Comment by felixmosh** [comment]
@madjam002, looks like the custom attribute feature is not suitable for that job, pls don't use it for that.

**Pull request closed without merging**